File: _2_2_Merge_Sort.py
def Merge(mergeIt, p, q, r):
    print("Merging")
    n1 = q-p+1
    n2 = r-p
    leftList = mergeIt[p:n1+1]
    rightList = mergeIt[q:n2+1]

    leftList[n1] = 10000
    rightList[n2] = 10000
    i = 1
    j = 1

    for num in range(p,r):
        if leftList[i] <= rightList[j]:
            mergeIt[num] = leftList[i]

            i = i + 1
        else:
            mergeIt[num] = rightList[j]
            j = j + 1





def MergeSort(listToSort, p, r):
    print("Inside MergeSort")
    if p < r:
       q = int((r - p) / 2)
       print(q)
       print(listToSort)
       MergeSort(listToSort, p, q)
       MergeSort(listToSort, q+1, r)
       Merge(listToSort, p, q, r)

listToSort = [9, 8, 7, 6, 5, 4, 3, 2, 1, 0]
r = len(listToSort)
print("size of the list: ",r)
p = 0
MergeSort(listToSort, p, r)

File: _2_2_Merge_Sort.c
#include<stdio.h>
int i = 0;
void merge_sort(int array[], int firstIndex, int lastIndex){
    //printf("\n%d inside mergesort",i);
    i++;
    //printf("%d, %d",firstIndex, lastIndex);
    if(firstIndex < lastIndex){
        printf("\ninside if");
        int midIndex = (firstIndex + lastIndex) / 2;
        printf("\nMid: %d, first: %d, last: %d",midIndex,firstIndex, lastIndex);
        printf("\nCalling 1st mergesort");
        merge_sort(array, firstIndex, midIndex);
        printf("\nCalling 2nd mergesort");
        merge_sort(array, midIndex+1, lastIndex);
        printf("\nCalling MERGE");
        merge(array, firstIndex, midIndex, lastIndex);
    }

}


void merge(int array[], int firstIndex, int midIndex, int lastIndex){
    printf("\ninside merge");
    int leftSize = midIndex - firstIndex + 1;
    int rightSize = lastIndex - midIndex;

    int leftArray[leftSize];
    int rightArray[rightSize];

    int i, j, k = 0;

    for(i = 0; i < leftSize; i++){
        leftArray[i] = array[firstIndex + i];
    }

    for(j = 0; j < rightSize; j++){
        rightArray[j] = array[midIndex + j + 1];
    }
    leftArray[leftSize] = 999999;
    rightArray[rightSize] = 999999;

    i = 0;
    j = 0;
    for(k = firstIndex; k <= lastIndex; k++){
        if(leftArray[i] <= rightArray[j]){
            array[k] = leftArray[i];
            i++;
        }
        else{
            array[k] = rightArray[j];
            j++;
        }
    }
    printf("\nThe merged array is: ");
    for(k = firstIndex; k <= lastIndex; k++){
        printf("%d, ", array[k]);
    }

}


void main(){
    int size = 10;
    int i = 0;
    printf("Array size: ");
    scanf("%d", &size);


    int array[10] = {10,9,8,7,6,5,1,2,3,4};
    printf("Array inputs: ");
    /*for(i = 0; i < size; i++){
        scanf("%d",&array[i]);
    }*/
    printf("Input done");
    merge_sort(array, 0, size-1);

}
